crypto/src/lib.rs

```diff
@@ -376,79 +501,132 @@ impl PublicKey {
+            }
+        }
     }
 }
 
 ffi::ffi_item! {
     /// Private Key used in signatures.
-    #[derive(Clone, PartialEq, Eq, CopyGetters)]
-    #[cfg_attr(not(feature="ffi_import"), derive(DebugCustom, Display, Serialize))]
-    #[cfg_attr(not(feature="ffi_import"), debug(fmt = "{{digest: {digest_function}, payload: {}}}", "hex::encode_upper(payload)"))]
-    #[cfg_attr(not(feature="ffi_import"), display(fmt = "{}", "hex::encode_upper(payload)"))]
-    pub struct PrivateKey {
-        /// Digest function
-        #[getset(get_copy = "pub")]
-        digest_function: Algorithm,
-        /// Key payload
-        #[serde(with = "hex::serde")]
-        payload: ConstVec<u8>,
+    #[derive(Clone)]
+    #[cfg_attr(all(feature = "ffi_export", not(feature = "ffi_import")), ffi_type(opaque))]
+    #[allow(missing_docs)]
+    pub enum PrivateKey {
+        Ed25519(Box<ed25519::PrivateKey>),
+        Secp256k1(secp256k1::PrivateKey),
+        BlsNormal(bls::BlsNormalPrivateKey),
+        BlsSmall(bls::BlsSmallPrivateKey),
     }
 }
 
-#[ffi_impl_opaque]
-impl PrivateKey {
-    /// Key payload
-    // TODO: Derive with getset once FFI impl is fixed
-    pub fn payload(&self) -> &[u8] {
-        self.payload.as_ref()
+impl PartialEq for PrivateKey {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Ed25519(l), Self::Ed25519(r)) => l == r,
+            (Self::Secp256k1(l), Self::Secp256k1(r)) => l == r,
+            (Self::BlsNormal(l), Self::BlsNormal(r)) => l.to_bytes() == r.to_bytes(),
+            (Self::BlsSmall(l), Self::BlsSmall(r)) => l.to_bytes() == r.to_bytes(),
+            _ => false,
+        }
     }
 }
 
+impl Eq for PrivateKey {}
+
 impl PrivateKey {
-    /// Construct `PrivateKey` from hex encoded string without validating the key
+    /// Creates a new public key from raw bytes received from elsewhere
     ///
     /// # Errors
     ///
-    /// If the given payload is not hex encoded
-    pub fn from_hex_unchecked(
-        digest_function: Algorithm,
-        payload: &(impl AsRef<[u8]> + ?Sized),
-    ) -> Result<Self, Error> {
-        Ok(Self {
-            digest_function,
-            payload: crate::hex_decode(payload).map(ConstVec::new)?,
-        })
+    /// - If the given payload is not a valid private key for the given digest function
+    pub fn from_raw(algorithm: Algorithm, payload: &[u8]) -> Result<Self, ParseError> {
+        match algorithm {
+            Algorithm::Ed25519 => ed25519::Ed25519Sha512::parse_private_key(payload)
+                .map(Box::new)
+                .map(Self::Ed25519),
+            Algorithm::Secp256k1 => {
+                secp256k1::EcdsaSecp256k1Sha256::parse_private_key(payload).map(Self::Secp256k1)
+            }
+            Algorithm::BlsNormal => bls::BlsNormal::parse_private_key(payload).map(Self::BlsNormal),
+            Algorithm::BlsSmall => bls::BlsSmall::parse_private_key(payload).map(Self::BlsSmall),
+        }
```

Comment:
the thing I mentioned before is that we don't use the `algorithm` value inside this function. This is an indication it could be a generic parameter like so:

```suggestion
    pub fn from_raw<T: Algorithm>(payload: &[u8]) -> Result<Self, ParseError> {
        T::parse_private_key(payload)
    }
```

not saying that we should do it like this. Just a food for thought

Reply:
since there are no const generics of enum variants this can only work if we make `Algorithm` a trait and it's variants into empty enums

Reply:
I get what you mean, but still there should be an entrypoint for user to *just* use `PrivateKey` key with algorithm got externally without hassle.